src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

```diff
@@ -136,4 +260,87 @@ RemoteQueryExecutor::Extension StorageObjectStorageCluster::getTaskIteratorExten
     return RemoteQueryExecutor::Extension{ .task_iterator = std::move(callback) };
 }
 
+std::shared_ptr<StorageObjectStorage> StorageObjectStorageCluster::getPureStorage(ContextPtr context)
+{
+    std::lock_guard lock(mutex);
+    if (!pure_storage)
+    {
+        pure_storage = std::make_shared<StorageObjectStorage>(
+            configuration,
+            object_storage,
+            context,
+            getStorageID(),
+            getInMemoryMetadata().getColumns(),
+            getInMemoryMetadata().getConstraints(),
+            comment,
+            format_settings,
+            mode,
+            /* distributed_processing */false,
+            partition_by);
+
+        auto virtuals_ = getVirtualsPtr();
+        if (virtuals_)
+            pure_storage->setVirtuals(*virtuals_);
+
+        pure_storage->setInMemoryMetadata(getInMemoryMetadata());
+    }
+
+    return pure_storage;
+}
+
+void StorageObjectStorageCluster::readFallBackToPure(
+    QueryPlan & query_plan,
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum processed_stage,
+    size_t max_block_size,
+    size_t num_streams)
+{
+    getPureStorage(context)->read(query_plan, column_names, storage_snapshot, query_info, context, processed_stage, max_block_size, num_streams);
+}
+
+SinkToStoragePtr StorageObjectStorageCluster::writeFallBackToPure(
+    const ASTPtr & query,
+    const StorageMetadataPtr & metadata_snapshot,
+    ContextPtr context,
+    bool async_insert)
+{
+    return getPureStorage(context)->write(query, metadata_snapshot, context, async_insert);
+}
+
+String StorageObjectStorageCluster::getClusterName(ContextPtr context) const
+{
+    auto cluster_name_ = context->getSettingsRef()[Setting::object_storage_cluster].value;
+    if (cluster_name_.empty())
+        cluster_name_ = getOriginalClusterName();
+    return cluster_name_;
+}
+
+QueryProcessingStage::Enum StorageObjectStorageCluster::getQueryProcessingStage(
+    ContextPtr context, QueryProcessingStage::Enum to_stage, const StorageSnapshotPtr & storage_snapshot, SelectQueryInfo & query_info) const
+{
+    /// Full query if fall back to pure storage.
+    if (getClusterName(context).empty())
+        return QueryProcessingStage::Enum::FetchColumns;
+
+    /// Distributed storage.
+    return IStorageCluster::getQueryProcessingStage(context, to_stage, storage_snapshot, query_info);
+}
+
+void StorageObjectStorageCluster::truncate(
+    const ASTPtr & query,
+    const StorageMetadataPtr & metadata_snapshot,
+    ContextPtr local_context,
+    TableExclusiveLockHolder & lock_holder)
+{
+    return getPureStorage(local_context)->truncate(query, metadata_snapshot, local_context, lock_holder);
```

Comment:
What happens if this method is called for a clustered instance?

Reply:
Shouldn't you do the same you did for `truncate`?

Reply:
Hm, may be. For cluster variants now disabled data modifications, I'll add check here too.